mtr.8
.Dd $Mdocdate$
.Dt MTR 8
.Os
.Sh NAME
.Nm mtr
.Nd a network diagnostic tool
.Sh SYNOPSIS
.Nm
.Op Fl abBcdefFgimnopPqrsStuvwxyzZ46
HOSTNAME ...
.Sh DESCRIPTION
.Nm
combines the functionality of the
.Sy traceroute
and
.Sy ping
programs in a single network diagnostic tool.
.Pp
As
.Sy Nm
starts, it investigates the network connection between the host
.Sy Nm
runs on and
.Sy Ns HOSTNAME
by sending packets with purposely low TTLs.  It continues to send packets with low TTL, noting the response time of the intervening routers.  This allows
.Sy Nm
to print the response percentage and response times of the internet route to
.Sy Ns HOSTNAME.
A sudden increase in packet loss or response time is often an indication of a bad (or simply overloaded) link.
.Pp
The results are usually reported as round-trip-response times in miliseconds and the percentage of packetloss.
.Sh OPTIONS

.Bl -tag -width Ds
.It Fl a, Fl -address Ar IP.ADD.RE.SS
Use this option to bind outgoing packets' socket to specific interface, so that any packet will be sent through this interface.
.Sy Note
that this option doesn't apply to DNS requests (which could be and could not be what you want).
.It Fl b, Fl -show-ips
Use this option to tell
.Sy Nm
to display both the host names and numeric IP numbers.  In split mode this adds an extra field to the output.  In report mode, there is usually too little space to add the IPs, and they will be truncated.  Use the wide report mode
.Fl w
to see the IPs in report mode.
.It Fl B, Fl -bitpattern Ar NUM
Specifies bit pattern to use in payload.
.br
Should be within range 0 - 255.
.It Fl c, Fl -cycles Ar COUNT
Use this option to set the number of pings sent to determine both the machines on the network and the reliability of those machines.  Each cycle lasts one second.
.It Fl d, Fl -display Ar MODE
Use this option to select the initial display mode
.Bl -tag -offset Ds
.It Cm 1st and 2nd bits
.br
.Sy 0
default
.br
.Sy 1
without latency information
.br
.Sy 2
with latency information
.br
.Sy 3
unicode histogram
.El
.Bl -tag -offset Ds
.It Cm 3rd bit
reserved
.It Cm 4th bit
color/mono mode, on/off
.It Cm 5th bit
audible bell (beep), on/off
.It Cm 6th bit
visible bell (flash), on/off
.It Cm 7th bit
bell for target host only, on/off
.El
.It Fl e, Fl -mpls
Use this option to tell
.Sy Nm
to display information from ICMP extensions for MPLS (RFC 4950) that are encoded in the response packets
.It Fl f, Fl -first-ttl Ar NUM
Specifies with what TTL to start.  Default is 1, `a' means auto (corresponding to TTL of the destination host).
.It Fl F, Fl -fields Ar DISPLAY-FIELDS
Use this option to specify the fields to display and their order
.Pp
Available fields:
.TS
allbox tab(%);
ll.
L%Loss ratio
D%Dropped packets
R%Received packets
S%Sent Packets
N%Newest RTT(ms)
B%Min/Best RTT(ms)
A%Average RTT(ms)
W%Max/Worst RTT(ms)
V%Standard Deviation
G%Geometric Mean
J%Current Jitter
M%Jitter Mean/Avg.
X%Worst Jitter
I%Interarrival Jitter
.TE

Examples:
.br
.Cm -F 'LSD NBAW'
.br
.Cm -F 'DR AGJMXI'
.It Fl g, Fl -graph Ar ARGUMENTS
Use this option to use the cairo based graph drawing (if available)
.Pp
Available arguments by order:
.Bl -tag -offset Ds
.It Cm 1st arg: graph type
1=dot 2=line 3=curve (default 3, curve)
.It Cm 2nd arg: viewport period
in 10sec ticks (default 6, i.e. 60sec)
.It Cm 3rd arg: legend on/off
0=disable 1=enable (default 1)
.It Cm 4th arg: multipath on/off
0=disable 1=enable (default 1)
.It Cm 5th arg: latency or jitter graph
0=latency 1=jitter (default 0, latency)
.El

Use `,' (comma) as argument separator

Examples:
.Bl -tag -width indent-two -offset Ds -compact
.It Cm -g,
all args in default (curve, 1min, legend, multipath, latency)
.It Cm -g2
line, default, default, default, default
.It Cm -g,30,,0
default, 5min, default, disable, default
.El
.It Fl i, Fl -interval Ar SECONDS
Use this option to specify the positive number of seconds between ICMP ECHO requests.  The default value for this parameter is one second.  The root user may choose values between zero and one.
.It Fl m, Fl -max-ttl Ar NUM
Specifies the maximum number of hops (max time-to-live value) traceroute will probe.  Default is 30.
.It Fl n, Fl -no-dns
Use this option to force
.Sy Nm
to display numeric IP numbers and not try to resolve the host names
.It Fl o, Fl -output Ar raw,txt,csv,json,xml
.RS
Use these suboptions for automated processing
.It Fl p, Fl -split
Use this option to set
.Sy Nm
to spit out a format that is suitable for a split-user interface
.It Fl P, Fl -port Ar PORT
The target port number for TCP traces
.It Fl Q, Fl -tos Ar NUM
Specifies value for type of service field in IP header.
.br
Should be within range 0 - 255.
.It Fl r, Fl -report
This option puts
.Sy Nm
into report mode.  When in this mode,
.Sy Nm
will run for the number of cycles specified by the
.Fl c
option, and then print statistics and exit.

This mode is useful for generating statistics about network quality.

Note that each running instance of
.Sy Nm
generates a significant amount of network traffic.  Using
.Sy Nm
to measure the quality of your network may result in decreased network performance.
.It Fl s, Fl -psize Ar BYTES
This option sets the packet size used for probing.  It is in bytes inclusive IP and ICMP headers.

If set to a negative number, every iteration will use a different, random packet size upto that number.
.It Fl S, Fl -summary
Print send/receive summary at exit
.It Fl t, Fl -tcp
Use TCP SYN packets instead of ICMP ECHO
.It Fl u, Fl -udp
Use UDP datagrams instead of ICMP ECHO
.It Fl v, Fl -version
.br
Print the installed version of
.Nm
.It Fl w, Fl -wide
This option puts
.Sy Nm
into wide report mode.  When in this mode,
.Sy Nm
will not cut hostnames in the report.
.It Fl x, Fl -cache Ar SECONDS
Cache mode. Don't ping known hops during cache-timeout period (0 means default 60 seconds).
.It Fl y, Fl -ipinfo Ar ORIGIN,FIELDS
Use this option to specify the IP info origin and its fields
.Pp
Available origins and their fields:
.Bl -tag -width indent -offset Ds
.It Cm 1 - origin.asn.cymru.com (both IPv4 and IPv6)
ASN, Route, CC, Registry, Allocated
.It Cm 2 - riswhois.ripe.net (both IPv4 and IPv6)
Route, Origin, Descr, CC
.It Cm 3 - peer.asn.shadowserver.org
Peers, ASN, Route, AS Name, CC, Org
.It Cm 4 - origin.asn.spameatingmonkey.net
Route, ASN, Org, Allocated, CC
.It Cm 5 - ip-api.com
Country, CC, RC, Region, City, Zip, Lat, Long, TZ, ISP, Org, AS Name
.It Cm 6 - asn.routeviews.org
ASN
.El

Abbreviations:
.Bl -tag -width indent-two -offset Ds -compact
.It Cm TZ
TimeZone
.It Cm Org
Organization
.It Cm DMA
Designated Market Area
.It Cm CC/RC/MC/LC
Country/Region/Metro/Local Code
.El
Examples:
.Bl -tag -width indent-two -offset Ds -compact
.It Cm -y5,2,5,7,8
ip-api.com: CC City Lat Long
.It Cm -y2,4,2
riswhois.ripe.net: CC Origin
.It Cm -y,
origin.asn.cymru.com: ASN
.El
.It Fl z, Fl -aslookup
Turn on ASN lookups. The data source is
.Lk riswhois.ripe.net
.It Fl Z, Fl -timeout Ar SECONDS
The number of seconds to keep the TCP socket open before giving up on the connection.  This will only affect the final hop.  Using large values for this, especially combined with a short interval, will use up a lot of file descriptors.
.It Fl 4
Use IPv4 only
.It Fl 6
Use IPv6 only

.Sh BUGS
Some modern routers give a lower priority to ICMP ECHO packets than to other network traffic.  Consequently, the reliability of these routers reported by
.Sy Nm
will be significantly lower than the actual reliability of these routers.
.Sh CONTACT INFORMATION
.Pp
.Pp
For more information about
.Sy Ns Nm
diagnostic tool, see web page at
.Lk http://www.bitwizard.nl/mtr
.Pp
Project's location is
.Lk https://github.com/traviscross/mtr
.Pp
This version is built from
.Lk https://github.com/yvs2014/mtr085
fork
.Sh "SEE ALSO"
traceroute(8),
ping(8),
TCP/IP Illustrated (Stevens, ISBN 0201633469)


mtr.c
/*
    mtr  --  a network diagnostic tool
    Copyright (C) 1997,1998  Matt Kimball

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 2 as
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <assert.h>
#include <fcntl.h>
#include <limits.h>
#include <sys/stat.h>

#include "config.h"

#ifdef HAVE_LIBIDN2
#include <idn2.h>
#define IDN_TO_ASCII_LZ	idn2_to_ascii_lz
#define IDN_TO_ASCII_8Z	idn2_to_ascii_8z
#elif HAVE_LIBIDN
#include <idna.h>
#define IDN_TO_ASCII_LZ	idna_to_ascii_lz
#define IDN_TO_ASCII_8Z	idna_to_ascii_8z
#endif

#ifdef UNICODE
#ifdef HAVE_WCTYPE_H
#include <wctype.h>
#endif
#ifdef HAVE_LOCALE_H
#include <locale.h>
#endif
#ifdef HAVE_LANGINFO_H
#include <langinfo.h>
#endif
#endif

#include "mtr.h"
#ifdef CURSES
#include "mtr-curses.h"
#endif
#include "getopt.h"
#include "display.h"
#include "dns.h"
#include "report.h"
#include "net.h"
#ifdef IPINFO
#include "ipinfo.h"
#endif
#ifdef GRAPHCAIRO
#include "graphcairo-mtr.h"
#endif
#include "macros.h"

#define CACHE_TIMEOUT 60

typedef struct names { char *name; struct names *next; } names_t;

// global vars
pid_t mypid;
char mtr_args[1024];	// posix/susvn: 4096+
unsigned iargs;		// args passed interactively
bool show_ips = false;
bool enable_mpls = false;
bool report_wide = false;
bool endpoint_mode = false; // -fa option, i.e. auto, corresponding to TTL of the destination host
bool cache_mode = false;            // don't ping known hops
int cache_timeout = CACHE_TIMEOUT;  // cache timeout in seconds
bool enable_stat_at_exit = false;
int fstTTL = 1;		// default start at first hop
int maxTTL = 30;	// enough?
int bitpattern;	// packet bit pattern used by ping
int remoteport = -1;	// target port
int tos;	// type of service set in ping packet
int cpacketsize = 64;	// default packet size
int tcp_timeout = 10 * 1000000;	// for TCP tracing
//
int display_offset;
int curses_mode;	// 1st and 2nd bits
#if defined(CURSES) || defined(GRAPHCAIRO)
int curses_mode_max = 3;
#endif
int color_mode;	// 4th bit
int audible_bell;	// 5th bit
int visible_bell;	// 6th bit
int target_bell_only;	// 7th bit
int mtrtype = IPPROTO_ICMP;	// Use ICMP as default packet type
	// default display field and order
int fld_index[256];
char fld_avail[AVLFLD + 1];
FLD_BUF_T fld_active;
FLD_BUF_T fld_save;
const struct fields data_fields[AVLFLD + 1] = {
	// Key, Remark, Header, Format, Width
	{' ', "<sp>: Space between fields", " ",  " ",        1},
	{'L', "L: Loss Ratio",          "Loss",   " %4.1f%%", 6},
	{'D', "D: Dropped Packets",     "Drop",   " %4d",     5},
	{'R', "R: Received Packets",    "Rcv",    " %5d",     6},
	{'S', "S: Sent Packets",        "Snt",    " %5d",     6},
	{'N', "N: Newest RTT(ms)",      "Last",   " %5.1f",   6},
	{'B', "B: Min/Best RTT(ms)",    "Best",   " %5.1f",   6},
	{'A', "A: Average RTT(ms)",     "Avg",    " %5.1f",   6},
	{'W', "W: Max/Worst RTT(ms)",   "Wrst",   " %5.1f",   6},
	{'V', "V: Standard Deviation",  "StDev",  " %5.1f",   6},
	{'G', "G: Geometric Mean",      "Gmean",  " %5.1f",   6},
	{'J', "J: Current Jitter",      "Jttr",   " %4.1f",   5},
	{'M', "M: Jitter Mean/Avg.",    "Javg",   " %4.1f",   5},
	{'X', "X: Worst Jitter",        "Jmax",   " %4.1f",   5},
	{'I', "I: Interarrival Jitter", "Jint",   " %4.1f",   5},
	{'\0', NULL, NULL, NULL, 0}
};

#if defined(OUTPUT_FORMAT_RAW) || defined(OUTPUT_FORMAT_TXT) || defined(OUTPUT_FORMAT_CSV) || defined(OUTPUT_FORMAT_JSON) || defined(OUTPUT_FORMAT_XML)
#define OUTPUT_FORMAT
#endif

static struct option long_options[] = {
  // Long, HasArgs, Flag, Short
  { "inet",       0, 0, '4' },  // use IPv4
#ifdef ENABLE_IPV6
  { "inet6",      0, 0, '6' },  // use IPv6
#endif
  { "address",    1, 0, 'a' },
  { "show-ips",   0, 0, 'b' },
  { "bitpattern", 1, 0, 'B' },   // overload b>255, ->rand(0,255)
  { "cycles",     1, 0, 'c' },
#if defined(CURSES) || defined(GRAPHCAIRO)
  { "display",    1, 0, 'd' },
#endif
  { "mpls",       0, 0, 'e' },
  { "first-ttl",  1, 0, 'f' },   // -f and -m are borrowed from traceroute
  { "fields",     1, 0, 'F' },   // fields to display and their order
#ifdef GRAPHCAIRO
  { "graph",      1, 0, 'g' },
#endif
  { "help",       0, 0, 'h' },
  { "interval",   1, 0, 'i' },
  { "max-ttl",    1, 0, 'm' },
  { "no-dns",     0, 0, 'n' },
#ifdef OUTPUT_FORMAT
  { "output",     1, 0, 'o' },  // output format: raw, txt, csv, json, xml
#endif
#ifdef SPLITMODE
  { "split",      0, 0, 'p' },
#endif
  { "port",       1, 0, 'P' },  // target port number for TCP
  { "tos",        1, 0, 'q' },  // typeof service (0..255)
  { "report",     0, 0, 'r' },
  { "psize",      1, 0, 's' },  // packet size
  { "summary",    0, 0, 'S' },  // print send/recv summary at exit
  { "tcp",        0, 0, 't' },  // TCP (default is ICMP)
  { "udp",        0, 0, 'u' },  // UDP (default is ICMP)
  { "version",    0, 0, 'v' },
  { "wide",       0, 0, 'w' },  // wide report (-r)
  { "cache",      1, 0, 'x' },  // enable cache with timeout in seconds (0 means default 60sec)
#ifdef IPINFO
  { "ipinfo",     1, 0, 'y' },
  { "aslookup",   0, 0, 'z' },
#endif
  { "timeout",    1, 0, 'Z' },  // timeout for TCP sockets
  { 0, 0, 0, 0 }
};
static char *short_options;

char srchost[NAMELEN];
char *dsthost = NULL;
int display_mode;
float wait_time = 1.0;
bool interactive = true;
int max_ping = 10;
bool alter_ping = false;
//

static char *iface_addr = NULL;
static names_t *names = NULL;

void set_fld_active(const char *s) {
  static FLD_BUF_T s_copy;
  strncpy((char*)s_copy, s, sizeof(s_copy) - 1);
  memset(fld_save, 0, sizeof(fld_save));
  strncpy((char*)fld_save, (char*)fld_active, sizeof(fld_save));
  memset(fld_active, 0, sizeof(fld_active));
  strncpy((char*)fld_active, (char*)s_copy, sizeof(fld_active));
}

char* trim(char *s) {
  char *p = s;
  int l = strlen(p);
  while (l && isspace((int)p[l-1])) // order matters
    p[--l] = 0;
  while (isspace((int)*p))
    p++;
  return p;
}

static void append_to_names(const char* item) {
  names_t* name = malloc(sizeof(names_t));
  if (name) {
    char* itemname = strdup(item);
    if (itemname) {
      name->name = itemname;
      name->next = names;
      names = name;
    } else {
      WARN_("strdup(%s)", item);
      free(name);
    }
  } else
    WARN_("malloc(%zd)", sizeof(names_t));
}

/*
 * If the file stream is associated with a regular file, lock/unlock the file
 * in order coordinate writes to a common file from multiple mtr
 * instances. This is useful if, for example, multiple mtr instances
 * try to append results to a common file.
 */
static void locker(FILE *file, short type) {
  if (!file)
    return;
  int fd = fileno(file);
  struct stat buf;
  if (fstat(fd, &buf) < 0) {
    WARN_("fstat(%d)", fd);
    return;
  }
  if (!S_ISREG(buf.st_mode))
    return;
  static struct flock l = { .l_whence = SEEK_END };
  l.l_type = type;
  l.l_pid = mypid;
  if (fcntl(fd, F_SETLKW, &l) < 0)
    WARN_("fcntl(fd=%d, type=%d)", fd, type);
}

static int my_getopt_long(int argc, char *argv[], int *opt_ndx) {
  if (!short_options) {
    short_options = calloc((sizeof(long_options) / sizeof(long_options[0])) * 2 + 1, 1);
    if (!short_options)
      return -1;
    char *p = short_options;
    for (int i = 0; long_options[i].name; i++) {
      *p++ = (char)long_options[i].val;
      if (long_options[i].has_arg)
        *p++ = ':';
    }
  }
  return getopt_long(argc, argv, short_options, long_options, opt_ndx);
}

static char *get_opt_desc(char opt) {
  switch (opt) {
    case 'm':
    case 'f':
    case 'B':
    case 'q':
    case 'P': return "NUMBER";
    case 'i':
    case 'x':
    case 'Z': return "SECONDS";
    case 'a': return "IP.ADD.RE.SS";
    case 'c': return "COUNT";
    case 'd': return "MODE";
    case 's': return "BYTES";
    case 'F': return "FIELDS";
#ifdef OUTPUT_FORMAT
    case 'o': return trim(""
#ifdef OUTPUT_FORMAT_RAW
" RAW"
#endif
#ifdef OUTPUT_FORMAT_TXT
" TXT"
#endif
#ifdef OUTPUT_FORMAT_CSV
" CSV"
#endif
#ifdef OUTPUT_FORMAT_JSON
" JSON"
#endif
#ifdef OUTPUT_FORMAT_XML
" XML"
#endif
);
#endif
#ifdef IPINFO
    case 'y': return "ORIGIN,FIELDS";
#endif
#ifdef GRAPHCAIRO
    case 'g': return "type,period,legend,multipath,jitter";
#endif
  }
  return NULL;
}

static void usage(char *name) {
  printf("Usage: %s [-", name);
  int l = strlen(short_options);
  for (int i = 0; i < l; i++)
    if (short_options[i] != ':')
      putchar(short_options[i]);
  printf("] HOSTNAME ...\n");
  for (int i = 0; long_options[i].name; i++) {
    printf("\t[");
    char c = (char)long_options[i].val;
    if (c)
      printf("-%c|", c);
    printf("--%s", long_options[i].name);
//    if (!c) // LATER: do it in more common way
//      if (!strncmp(long_options[i].name, SOME_LONG_OPTION, sizeof(SOME_LOG_OPTION)))
//        c = 'i'; // seconds
    char *desc = long_options[i].has_arg ? get_opt_desc(c) : NULL;
    if (desc)
      printf(" %s", desc);
    printf("]\n");
  }
}

static int limit_int(const int v0, const int v1, const int v, const char *it) {
  if (v < v0) {
    WARNX_("'%s' is out of range %d..%d: %d -> %d", it, v0, v1, v, v0);
    return v0;
  }
  if (v > v1) {
    WARNX_("'%s' is out of range %d..%d: %d -> %d", it, v0, v1, v, v1);
    return v1;
  }
  return v;
}
 
static void parse_options(int argc, char **argv) {
  while (1) {
    int opt = my_getopt_long(argc, argv, NULL);
    if (opt == -1)
      break;

    switch (opt) {
    case '4':
      net_init(0);
      break;
#ifdef ENABLE_IPV6
    case '6':
      net_init(1);
      break;
#endif
    case '?':
      usage(argv[0]);
      exit(EXIT_FAILURE);
    case 'a':
      iface_addr = optarg;
      break;
    case 'b':
      show_ips = true;
      break;
    case 'B':
      bitpattern = atoi(optarg);
      if (bitpattern > 255)
        bitpattern = -1;
      break;
    case 'c':
      max_ping = atoi(optarg);
      alter_ping = true;
      break;
#if defined(CURSES) || defined(GRAPHCAIRO)
    case 'd':
      curses_mode = ((atoi(optarg)) & ~8) % curses_mode_max;
      color_mode = ((atoi(optarg)) & 8) ? 1 : 0;
      audible_bell = ((atoi(optarg)) & 16) ? 1 : 0;
      visible_bell = ((atoi(optarg)) & 32) ? 1 : 0;
      target_bell_only = ((atoi(optarg)) & 64) ? 1 : 0;
      break;
#endif
    case 'e':
      enable_mpls = true;
      break;
    case 'f':
      if (optarg[0] == 'a') {
        endpoint_mode = true;
        break;
      }
      fstTTL = limit_int(1, maxTTL, atoi(optarg), "first ttl");
      break;
    case 'F':
      if (strlen(optarg) >= sizeof(fld_active))
        FAIL_("-F: Too many fields (max=%zd): %s", sizeof(fld_active) - 1, optarg);
      for (int i = 0; optarg[i]; i++)
        if(!strchr(fld_avail, optarg[i]))
          FAIL_("-F: Unknown field identifier %c", optarg[i]);
      set_fld_active(optarg);
      break;
#ifdef GRAPHCAIRO
    case 'g':
      gc_parsearg(optarg);
      display_mode = DisplayGraphCairo;
      break;
#endif
    case 'h':
      usage(argv[0]);
      exit(EXIT_SUCCESS);
    case 'i':
      wait_time = atof(optarg);
      if (wait_time <= 0)
        FAIL("-i: Wait time must be positive");
      if (getuid() && (wait_time < 1))
        FAIL("-i: Non-root users cannot set an interval less than one second");
      break;
    case 'm':
      maxTTL = limit_int(1, ((MAXHOST - 1) > maxTTL) ? maxTTL : (MAXHOST - 1), atoi(optarg), "maximum ttl");
      break;
    case 'n':
      enable_dns = false;
      break;
#ifdef OUTPUT_FORMAT
    case 'o':
      switch (tolower((int)optarg[0])) {
#ifdef OUTPUT_FORMAT_RAW
        case 'r':
          display_mode = DisplayRaw;
          enable_raw = true;
          break;
#endif
#ifdef OUTPUT_FORMAT_TXT
        case 't':
          display_mode = DisplayTXT;
          break;
#endif
#ifdef OUTPUT_FORMAT_CSV
        case 'c':
          display_mode = DisplayCSV;
          break;
#endif
#ifdef OUTPUT_FORMAT_JSON
        case 'j':
          display_mode = DisplayJSON;
          break;
#endif
#ifdef OUTPUT_FORMAT_XML
        case 'x':
          display_mode = DisplayXML;
          break;
#endif
        default:
          usage(argv[0]);
          exit(EXIT_FAILURE);
      }
      break;
#endif
#ifdef SPLITMODE
    case 'p':
      display_mode = DisplaySplit;
      break;
#endif
    case 'P':
      remoteport = atoi(optarg);
      if (remoteport > 65535 || remoteport < 1)
        FAIL_("-P: Illegal port number %d", remoteport);
      break;
    case 'q':
      tos = atoi(optarg);
      if (tos > 255 || tos < 0)	// error message, should do more checking for valid values
        tos = 0;
      break;
    case 'r':
      display_mode = DisplayReport;
      break;
    case 's':
      cpacketsize = atoi(optarg);
      break;
    case 'S':
      enable_stat_at_exit = true;
      break;
    case 't':
      if (mtrtype != IPPROTO_ICMP)
        FAIL("-t and -u are mutually exclusive");
      if (net_tcp_init())
        mtrtype = IPPROTO_TCP;
      else
        FAIL("-t: Switch to TCP mode failed");
      break;
    case 'u':
      if (mtrtype != IPPROTO_ICMP)
        FAIL("-u and -t are mutually exclusive");
      mtrtype = IPPROTO_UDP;
      break;
    case 'v':
      printf("%s-%s\n", argv[0], MTR_VERSION);
      exit(EXIT_SUCCESS);
    case 'w':
      report_wide = true;
      display_mode = DisplayReport;
      break;
    case 'x':
      cache_mode = true;
      cache_timeout = atoi(optarg);
	  if (cache_timeout < 0)
        FAIL_("-x: Cache timeout %d must be positive", cache_timeout);
      else if (cache_timeout == 0)
        cache_timeout = CACHE_TIMEOUT;  // default 60 seconds
      break;
#ifdef IPINFO
    case 'y':
      if (ipinfo_args(optarg) < 0)
        exit(EXIT_FAILURE);
      break;
    case 'z':
      ipinfo_args(ASLOOKUP_DEFAULT);
      break;
#endif
    case 'Z':
      tcp_timeout = atoi(optarg);
      tcp_timeout *= 1000000;
      break;
    default:
      usage(argv[0]);
      exit(EXIT_FAILURE);
    }
  }

  static int sz;
  for (int i = 1; i < argc; i++)
    sz += snprintf(mtr_args + sz, sizeof(mtr_args) - sz, " %s", argv[i]);

  switch (display_mode) {
    case DisplayReport:
#ifdef OUTPUT_FORMAT_RAW
    case DisplayRaw:
#endif
#ifdef OUTPUT_FORMAT_TXT
    case DisplayTXT:
#endif
#ifdef OUTPUT_FORMAT_CSV
    case DisplayCSV:
#endif
#ifdef OUTPUT_FORMAT_JSON
    case DisplayJSON:
#endif
#ifdef OUTPUT_FORMAT_XML
    case DisplayXML:
#endif
      interactive = false;
  }

  if (optind > argc - 1)
    return;
}


#define IDNA_RESOLV(func) { \
  rc = func(dsthost, &hostname, 0); \
  if (rc == IDNA_SUCCESS) rc = getaddrinfo(hostname, NULL, &hints, &res); \
  else rc_msg = idna_strerror(rc); \
  if (hostname) free(hostname); \
}


static bool set_host(struct addrinfo *res) {
  struct addrinfo *ai;
  for (ai = res; ai; ai = ai->ai_next)
    if (af == ai->ai_family)  // use only the desired AF
      break;
  if (af != ai->ai_family) {  // not found
    WARNX_("Desired address family %d not found", af);
    return false;
  }

  char* alptr[2] = { NULL, NULL };
  if (af == AF_INET)
    alptr[0] = (void*) &(((struct sockaddr_in *)ai->ai_addr)->sin_addr);
#ifdef ENABLE_IPV6
  else if (af == AF_INET6)
    alptr[0] = (void*) &(((struct sockaddr_in6 *)ai->ai_addr)->sin6_addr);
#endif
  else {
    WARNX_("Unknown address family %d", af);
    return false;
  }

  static struct hostent h;
  memset(&h, 0, sizeof(h));
  h.h_name = ai->ai_canonname;
  h.h_aliases = NULL;
  h.h_addrtype = ai->ai_family;
  h.h_length = ai->ai_addrlen;
  h.h_addr_list = alptr;

  if (!net_set_host(&h)) {
    WARNX("Unable to set host entry");
    return false;
  }
  if (!net_set_ifaddr(iface_addr)) {
    WARNX("Unable to set interface address");
    return false;
  }
  return true;
}


int main(int argc, char **argv) {
  if (!net_open())  // Get the raw sockets first thing, so we can drop to user euid immediately
    FAIL("Unable to get raw sockets");
  if (setgid(getgid()) || setuid(getuid()))  // Now drop to user permissions
    ERRR(EXIT_FAILURE, "Unable to drop permissions");
  if ((geteuid() != getuid()) || (getegid() != getgid())) // Double check, just in case
    FAIL("Unable to drop permissions");

  mypid = getpid();

  /* reset the random seed */
  srand(mypid);

  display_detect(&argc, &argv);
#if defined(CURSES) || defined(GRAPHCAIRO)
  curses_mode = 0;
#endif

  /* The field options are now in a static array all together,
     but that requires a run-time initialization. */
  memset(fld_index, -1, sizeof(fld_index));
  for (int i = 0; (data_fields[i].key) && (i < sizeof(fld_avail)); i++) {
    fld_avail[i] = data_fields[i].key;
    fld_index[data_fields[i].key] = i;
  }

#ifdef UNICODE
  bool dm_histogram = false;
#ifdef HAVE_LOCALE_H
  char *lc_ctype = setlocale(LC_CTYPE, NULL);
  setlocale(LC_CTYPE, "");
#ifdef HAVE_LANGINFO_H
  if (strcasecmp("UTF-8", nl_langinfo(CODESET)) == 0) {
    if (iswprint(L'▁'))
      dm_histogram = true;
    else
      WARNX("Unicode block elements are not printable");
  }
#endif
  if (dm_histogram)
    curses_mode_max++;
  else
    setlocale(LC_CTYPE, lc_ctype);
#endif
#endif

  set_fld_active(FLD_ACTIVE_DEFAULT);
  parse_options(argc, argv);

  while (optind < argc) {
    char* name = argv[optind++];
    append_to_names(name);
  }

  if (!names) {
    usage(argv[0]);
    exit(EXIT_SUCCESS);
  }

  /* Now that we know mtrtype we can select which socket to use */
  if (net_selectsocket() < 0)
    FAIL("Unable to determine type of socket");

  dns_open();
#ifdef IPINFO
  ipinfo_open();
#endif
  if (gethostname(srchost, sizeof(srchost)))
    strncpy(srchost, "UNKNOWN", sizeof(srchost));
  display_start();

  for (names_t* n = names; n; n = n->next) {
    static bool notfirst;
    dsthost = n->name;

    struct addrinfo hints, *res;
    memset(&hints, 0, sizeof(hints));
    hints.ai_family = af;
    hints.ai_socktype = SOCK_DGRAM;
    int rc = getaddrinfo(dsthost, NULL, &hints, &res);
    const char *rc_msg = NULL;
#if defined(HAVE_LIBIDN) || defined(HAVE_LIBIDN2)
    if (rc) {
      char *hostname = NULL;
      IDNA_RESOLV(IDN_TO_ASCII_LZ);
      if (rc)
        IDNA_RESOLV(IDN_TO_ASCII_8Z);
    }
#endif
    if (rc) {
      if (rc == EAI_SYSTEM)
        WARN_("getaddrinfo(%s)", dsthost);
      else
        WARNX_("Failed to resolve \"%s\": %s", dsthost, rc_msg ? rc_msg : gai_strerror(rc));
    } else {
      if (set_host(res)) {
        locker(stdout, F_WRLCK);
        if (display_open())
          display_loop();
        else
          WARNX("Unable to open display");
        net_end_transit();
        display_close(notfirst);
        locker(stdout, F_UNLCK);
      }
      freeaddrinfo(res);
    }
    notfirst = true;
  }

  display_finish();
#ifdef IPINFO
  ipinfo_close();
#endif
  dns_close();
  net_close();
  if (enable_stat_at_exit) {
    printf("NET: %lu queries (%lu icmp, %lu udp, %lu tcp), %lu replies (%lu icmp, %lu udp, %lu tcp)\n",
      net_queries[0], net_queries[1], net_queries[2], net_queries[3],
      net_replies[0], net_replies[1], net_replies[2], net_replies[3]);
    printf("DNS: %u queries (%u ptr, %u txt), %u replies (%u ptr, %u txt)\n",
      dns_queries[0], dns_queries[1], dns_queries[2],
      dns_replies[0], dns_replies[1], dns_replies[2]);
#ifdef IPINFO
    printf("EXTRA: %u queries (%u http, %u whois), %u replies (%u http, %u whois)\n",
      ipinfo_queries[0], ipinfo_queries[1], ipinfo_queries[2],
      ipinfo_replies[0], ipinfo_replies[1], ipinfo_replies[2]);
#endif
  }
  return 0;
}

